src/gen_pfc.c

```diff
@@ -241,68 +241,181 @@ static void _gen_pfc_chain(const struct arch_def *arch,
  *
  */
 static void _gen_pfc_syscall(const struct arch_def *arch,
-			     const struct db_sys_list *sys, FILE *fds)
+			     const struct db_sys_list *sys, FILE *fds,
+			     int lvl)
 {
 	unsigned int sys_num = sys->num;
 	const char *sys_name = arch_syscall_resolve_num(arch, sys_num);
 
-	_indent(fds, 1);
+	_indent(fds, lvl);
 	fprintf(fds, "# filter for syscall \"%s\" (%u) [priority: %d]\n",
 		(sys_name ? sys_name : "UNKNOWN"), sys_num, sys->priority);
-	_indent(fds, 1);
+	_indent(fds, lvl);
 	fprintf(fds, "if ($syscall == %u)\n", sys_num);
 	if (sys->chains == NULL) {
-		_indent(fds, 2);
+		_indent(fds, lvl + 1);
 		_pfc_action(fds, sys->action);
 	} else
-		_gen_pfc_chain(arch, sys->chains, 2, fds);
+		_gen_pfc_chain(arch, sys->chains, lvl + 1, fds);
 }
 
-/**
- * Generate pseudo filter code for an architecture
- * @param col the seccomp filter collection
- * @param db the single seccomp filter
- * @param fds the file stream to send the output
- *
- * This function generates a pseudo filter code representation of the given
- * filter DB and writes it to the given output stream.  Returns zero on
- * success, negative values on failure.
- *
- */
-static int _gen_pfc_arch(const struct db_filter_col *col,
-			 const struct db_filter *db, FILE *fds)
+#define SYSCALLS_PER_NODE		(4)
+#define MIN_SYSCALLS_TO_USE_BINTREE	(16)
+static int _get_bintree_levels(unsigned int syscall_cnt,
+			       uint32_t optimize)
 {
-	int rc = 0;
+	unsigned int i = 0, max_level;
+
+	if (syscall_cnt < MIN_SYSCALLS_TO_USE_BINTREE ||
+	    optimize != 2)
+		/* Only use a binary tree if there are a lot of syscalls
+		 * and if the filter optimization level is set to 2 (aka
+		 * binary tree)
+		 */
+		return 0;
+
+	do {
+		max_level = SYSCALLS_PER_NODE << i;
+		i++;
+	} while(max_level < syscall_cnt);
+
+	return i;
+}
+
+static int _get_bintree_syscall_num(const struct pfc_sys_list *cur,
+				    int lookahead_cnt,
+				    int * const num)
+{
+	while (lookahead_cnt > 0 && cur != NULL) {
+		cur = cur->next;
+		lookahead_cnt--;
+	}
+
+	if (cur == NULL)
+		return -EINVAL;
+
+	*num = cur->sys->num;
+	return 0;
+}
+
+static int _sort_syscalls_by_num(struct db_sys_list *syscalls,
+				 struct pfc_sys_list **p_head)
+{
+	struct pfc_sys_list *p_iter = NULL, *p_new, *p_prev;
 	struct db_sys_list *s_iter;
-	struct pfc_sys_list *p_iter = NULL, *p_new, *p_head = NULL, *p_prev;
 
-	/* sort the syscall list */
-	db_list_foreach(s_iter, db->syscalls) {
+	db_list_foreach(s_iter, syscalls) {
 		p_new = zmalloc(sizeof(*p_new));
 		if (p_new == NULL) {
-			rc = -ENOMEM;
-			goto arch_return;
+			return -ENOMEM;
 		}
 		p_new->sys = s_iter;
 
 		p_prev = NULL;
-		p_iter = p_head;
+		p_iter = *p_head;
+		while (p_iter != NULL &&
```

Comment:
The only thing I might say here is that having pfc_sys_list be an array of some type instead would enable you to use qsort() or something from the standard library. You'd have to switch to an array, re-allocing along the way, etc. to use it that way, but that might be less scary than sorting things manually with all the pointer math that's thrown in above and below here.

Alternatively, you could just copy the pointers into an array, and use qsort() with a comparator that does a deref. That might be less scary code too.

Reply:
That's an interesting idea.

I remember when I wrote this a long time back that I wanted to keep the `_sort_syscalls_by_num()` logic as close to the original `_sort_syscalls_by_priority()` logic as possible.  Seemed reasonable at the time.

Now looking at it a year later, I agree.  It's more cryptic than I would like.

Reply:
Yeah, definitely no need to block this PR on the above. But also it's a matter of taste, and it's working and tested the way it is, so :man_shrugging:

Reply:
Thanks.  Unless others object, I'm leaning that way as well.

But, I admit I may revisit this function in the future.  I have been wanting to add some unit testing to key parts of the code (some of the `db_*()` functions for example), and this may be a good place to start.  It has a fairly simple set of inputs and the desired results are also straightforward.

Reply:
FWIW, the gen_pfc code doesn't need to be *great*, it just needs to work.  It's there primarily as a debug tool, not something that would ever be used in production (it doesn't generate any *real* filter code after all).